src/value.c

```diff
@@ -327,29 +159,68 @@ size_t SIValue_StringConcat(SIValue* strings, unsigned int string_count, char *b
   return offset;
 }
 
-int SIValue_Compare(SIValue a, SIValue b) {
-  // In order to be comparable, both SIValues must be strings,
-  // booleans, or numerics
-  if (!SI_COMPARABLE(a, b)) {
-    return DISJOINT;
+SIValue SIValue_Add(const SIValue a, const SIValue b) {
+  /* Only construct an integer return if both operands are integers. */
+  if (a.type & b.type & T_INT64) {
+    return SI_LongVal(a.longval + b.longval);
   }
+  /* Return a double representation. */
+  return SI_DoubleVal(SI_GET_NUMERIC(a) + SI_GET_NUMERIC(b));
+}
 
-  // Use strcmp if values are string types
-  if (a.type & SI_STRING) return strcmp(a.stringval, b.stringval);
-
-  // Attempt to cast both values to doubles
-  double tmp_a, tmp_b;
-  if (SIValue_ToDouble(&a, &tmp_a) && SIValue_ToDouble(&b, &tmp_b)) {
-    /* Both values are numeric, and both now have double representations.
-     * TODO worry about precision and overflows. This approach will
-     * not be adequate if we have high-value longs, especially.
-     * TODO Special values like inf, -inf, and NaN will compare properly
-     * here, but may not satisfy the prescribed openCypher sort order. */
-    double diff = tmp_a - tmp_b;
+SIValue SIValue_Subtract(const SIValue a, const SIValue b) {
+  /* Only construct an integer return if both operands are integers. */
+  if (a.type & b.type & T_INT64) {
+    return SI_LongVal(a.longval - b.longval);
+  }
+  /* Return a double representation. */
+  return SI_DoubleVal(SI_GET_NUMERIC(a) - SI_GET_NUMERIC(b));
+}
+
+SIValue SIValue_Multiply(const SIValue a, const SIValue b) {
+  /* Only construct an integer return if both operands are integers. */
+  if (a.type & b.type & T_INT64) {
+    return SI_LongVal(a.longval * b.longval);
+  }
+  /* Return a double representation. */
+  return SI_DoubleVal(SI_GET_NUMERIC(a) * SI_GET_NUMERIC(b));
+}
+
+SIValue SIValue_Divide(const SIValue a, const SIValue b) {
+  /* Always perform floating-point division. */
+  return SI_DoubleVal(SI_GET_NUMERIC(a) / (double)SI_GET_NUMERIC(b));
+}
+
+int SIValue_Compare(const SIValue a, const SIValue b) {
+  /* In order to be comparable, both SIValues must be strings,
+   * booleans, or numerics. */
+  SIType t = a.type ^ b.type;
+  if (!t) {
+    // Both inputs have the same type
+    switch (a.type) {
+      case T_INT64:
+      case T_BOOL:
+        return a.longval - b.longval;
+      case T_DOUBLE:
+        return COMPARE_RETVAL(a.doubleval - b.doubleval);
+      case T_STRING:
+      case T_CONSTSTRING:
+        // Both inputs are strings of the same SIType
+        return strcmp(a.stringval, b.stringval);
+      default:
+        // Both pointers were of an incomparable type, like a pointer
+        return DISJOINT;
+    }
+  } else if (t == SI_NUMERIC) {
```

Comment:
This is confusing, t is a product of XOR, 
Without looking at the actual type enum values, it is hard to reason about the result of this xor being equal to a specific enum type. and so I think it will be better to explicitly specify the types here
if SI_STRING(A) && SI_STRING(B)...

Reply:
I concede that it's confusing! The trick here is that since we have two numeric types and two string types, if both inputs are numerics or both are strings, the XOR will resolve to 0 for exact matches and to the entirety of SI_NUMERIC or SI_STRING for differing but comparable types. Anything else is an invalid comparison.

This is a pretty heavily-used function, so I'm personally okay with cutting out a few comparisons at the expense of readability. I'll change it if you think it's too cryptic, though!